### PR TITLE
TF-3005 [SEARCH] Easily clear "from" and "to" search result header

### PR DESCRIPTION
**Pull request by dab246**
## Issue

#3005 

## Resolved

- Web:

https://github.com/user-attachments/assets/1441b5c3-228e-4897-a5a8-c01d834f174e

- Mobile:

[android.webm](https://github.com/user-attachments/assets/d53362da-51da-40bf-a8ac-f717392f2d77)

### CODE REVIEW

**Comment by hoangdat** [comment]
- [ ] IMO, we should display `tick` to indicated selected email address. Then hover will change to `delete` button 
image

File: lib/features/contact/presentation/contact_controller.dart [hunk 1]
```diff
@@ -162,14 +181,46 @@ class ContactController extends BaseController {
...
+    textInputSearchFocus.unfocus();
+    FocusManager.instance.primaryFocus?.unfocus();
+    popBack(result: selectedContactList);
+  }
+
+  void handleOnDoneAction() {
+    textInputSearchFocus.unfocus();
+    FocusManager.instance.primaryFocus?.unfocus();
+    popBack(result: selectedContactList);
+  }
```
Comment (hoangdat) [inline lib/features/contact/presentation/contact_controller.dart:216]: how about if click outside the dialog ?

File: lib/features/contact/presentation/widgets/contact_item_widget.dart [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+              onTapActionCallback: () =>
+                widget.onDeleteContactAction!(widget.emailAddress),
+            )
+          else if (PlatformInfo.isMobile && widget.onDeleteContactAction != null)
+            TMailButtonWidget.fromIcon(
+              icon: widget.imagePaths.icClearTextSearch,
+              backgroundColor: Colors.transparent,
+              onTapActionCallback: () =>
+                widget.onDeleteContactAction!(widget.emailAddress),
+            )
```
Comment (tddang-linagora) [inline lib/features/contact/presentation/widgets/contact_item_widget.dart:105]: Should combine these 2 `if` statements to
```dart
bool get canDeleteContact {
  // ...
}
```

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 3]
```diff
@@ -1696,38 +1696,48 @@ class MailboxDashBoardController extends ReloadableController with UserSettingPo
     dispatchAction(StartSearchEmailAction());
   }
 
-  Future<void> selectFromSearchFilter() async {
+  Future<void> selectFromSearchFilter(BuildContext context) async {
```
Comment (tddang-linagora) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1699]:
```suggestion
  Future<void> selectFromSearchFilter(AppLocalizations appLocalizations) async {
```

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 4]
```diff
@@ -1696,38 +1696,48 @@ class MailboxDashBoardController extends ReloadableController with UserSettingPo
...
     if (accountId.value == null || sessionCurrent == null) return;
 
     final contactArgument = ContactArguments(
-      accountId.value!,
-      sessionCurrent!,
-      searchController.searchEmailFilter.value.from);
+      accountId: accountId.value!,
+      session: sessionCurrent!,
+      selectedContactList: searchController.searchEmailFilter.value.from,
+      contactViewTitle: '${AppLocalizations.of(context).findEmails} ${AppLocalizations.of(context).from_email_address_prefix.toLowerCase()}'
```
Comment (tddang-linagora) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1706]:
```suggestion
      contactViewTitle: '${appLocalizations.findEmails} ${appLocalizations.from_email_address_prefix.toLowerCase()}'
```

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 5]
```diff
@@ -1696,38 +1696,48 @@ class MailboxDashBoardController extends ReloadableController with UserSettingPo
...
+      final listMailAddress = newListContact
+        .map((emailAddress) => emailAddress.emailAddress)
+        .toSet();
+      searchController.updateFilterEmail(fromOption: Some(listMailAddress));
       dispatchAction(StartSearchEmailBySearchFilterAction(QuickSearchFilter.from));
     }
   }
 
-  Future<void> selectToSearchFilter() async {
+  Future<void> selectToSearchFilter(BuildContext context) async {
```
Comment (tddang-linagora) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1722]: idem

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 6]
```diff
@@ -1696,38 +1696,48 @@ class MailboxDashBoardController extends ReloadableController with UserSettingPo
...
     if (accountId.value == null || sessionCurrent == null) return;
 
     final contactArgument = ContactArguments(
-      accountId.value!,
-      sessionCurrent!,
-      searchController.searchEmailFilter.value.to);
+      accountId: accountId.value!,
+      session: sessionCurrent!,
+      selectedContactList: searchController.searchEmailFilter.value.to,
+      contactViewTitle: '${AppLocalizations.of(context).findEmails} ${AppLocalizations.of(context).to_email_address_prefix.toLowerCase()}'
```
Comment (tddang-linagora) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1729]: idem

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart [hunk 7]
```diff
@@ -591,13 +591,13 @@ class MailboxDashBoardView extends BaseMailboxDashBoardView {
         }
         break;
       case QuickSearchFilter.from:
-        controller.selectFromSearchFilter();
+        controller.selectFromSearchFilter(context);
```
Comment (tddang-linagora) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:594]:
```suggestion
        controller.selectFromSearchFilter(AppLocalizations.of(context));
```

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart [hunk 8]
```diff
@@ -591,13 +591,13 @@ class MailboxDashBoardView extends BaseMailboxDashBoardView {
...
         break;
       case QuickSearchFilter.hasAttachment:
         controller.selectHasAttachmentSearchFilter();
         break;
       case QuickSearchFilter.to:
-        controller.selectToSearchFilter();
+        controller.selectToSearchFilter(context);
```
Comment (tddang-linagora) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:600]: idem

File: lib/features/contact/presentation/contact_controller.dart [hunk 9]
```diff
@@ -162,14 +181,46 @@ class ContactController extends BaseController {
...
 
-  void selectContact(BuildContext context, EmailAddress emailAddress) {
-    KeyboardUtils.hideKeyboard(context);
-    popBack(result: emailAddress);
+  void handleOnSelectContactAction(EmailAddress emailAddress) {
+    log('ContactController::selectContact:emailAddress = $emailAddress');
+    final isEmailAddressExist = selectedContactList
+      .any((contact) => contact.emailAddress == emailAddress.emailAddress);
+    log('ContactController::selectContact:isEmailAddressExist = $isEmailAddressExist');
+    if (!isEmailAddressExist) {
```
Comment (tddang-linagora) [inline lib/features/contact/presentation/contact_controller.dart:189]: If the contact exists and user select it, should we remove the contact out of `selected`?

File: lib/features/contact/presentation/contact_controller.dart (hunk 1 above)
Comment (dab246) [reply lib/features/contact/presentation/contact_controller.dart:216]: click outside disabled

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart (hunk 4 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1706]: why?

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart (hunk 5 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1722]: why?

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart (hunk 6 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1729]: why?

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart (hunk 7 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:594]: Why?

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart (hunk 8 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:600]: Why?

File: lib/features/contact/presentation/contact_controller.dart (hunk 9 above)
Comment (dab246) [reply lib/features/contact/presentation/contact_controller.dart:189]: No, we have a separate delete button for it.

**Comment by dab246** [comment]
> * [ ]  IMO, we should display `tick` to indicated selected email address. (same as mobile)
> 
> image

What do you mean? Describe in detail.

File: lib/features/contact/presentation/widgets/contact_item_widget.dart (hunk 2 above)
Comment (dab246) [reply lib/features/contact/presentation/widgets/contact_item_widget.dart]: Fixed

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart (hunk 3 above)
Comment (tddang-linagora) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1699]: Avoid passing around `BuildContext`, when all that method need is `AppLocalizations` from the `BuildContext`

Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:1699]: We will need more than just AppLocalizations from BuildContext

**Review by tddang-linagora — APPROVED** [review]

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart (hunk 8 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart]: Fixed

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart (hunk 7 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart]: Fixed

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart (hunk 4 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart]: Fixed

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart (hunk 3 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart]: Fixed

**Comment by dab246** [comment]
> > * [x]   IMO, we should display `tick` to indicated selected email address. (same as mobile)
> > 
> > image
> 
> What do you mean? Describe in detail.

Updated

- Web:


https://github.com/user-attachments/assets/2487446a-c9f5-4bac-9b1d-464edfb17ee3

- Mobile:

[Screen_recording_20241008_162823.webm](https://github.com/user-attachments/assets/7b640b2e-851d-469b-b468-625834b4ab73)

**Review by tddang-linagora — APPROVED** [review]

**Comment by hoangdat** [comment]
- [ ] if no selected email address, can we focus on search bar

https://github.com/user-attachments/assets/12aec49d-3076-4ccf-aa3e-3db475ca90c7

**Comment by hoangdat** [comment]
- [ ] should select text in search if user select a suggestion, it will make them easier to find the next one

image

**Comment by dab246** [comment]
> * [x]  if no selected email address, can we focus on search bar
> 
>  Screen.Recording.2024-10-08.at.17.35.48.mov

Fixed


https://github.com/user-attachments/assets/c462cd03-2a31-49c4-83e8-b7c5040706f0

**Comment by hoangdat** [comment]
- [ ] if we select multiple recipients, we should combine it as `OR` operation. WDYT?

image

**Comment by hoangdat** [comment]
- [ ] when app has realtime update, search result also reload. We should not reload it

**Comment by dab246** [comment]
> * [ ]  if we select multiple recipients, we should combine it as `OR` operation. WDYT?
> 
> image

It is absurd, because when user enters a field with multiple mail addresses, it only searches for emails that include all of these mail addresses. And the `AND` condition is reasonable.

**Review by hoangdat — APPROVED** [review]

**Pull request merged**